web/internal/metadata.bzl

```diff
@@ -0,0 +1,153 @@
+  if browser_label:
+    content += ',\n  "browserLabel": "%s"' % browser_label
+  if test_label:
+    content += ',\n  "testLabel": "%s"' % test_label
+  if web_test_files:
+    content += ',\n  "webTestFiles": ['
+    first = True
+    for f in web_test_files:
+      if first:
+        first = False
+      else:
+        content += ','
+      content += '\n' + _web_test_files_to_string(ctx, f, '    ')
+    content += '\n  ]'
+  content += '\n}\n'
+
+  ctx.file_action(output=output, content=content, executable=False)
+
+
+def _get_files(ctx, attr_names):
+  """Finds all of the web_test_metadata files in given attributes.
+
+  It ensures that all web_test_metadata files position in the returned list is
+  based on the last attribute in attr_names that they appear in.
+
+  Args:
+    ctx: the context object for this rule.
+    attr_names: the names of attributes to get web_test_metadata files from.
+  Returns:
+    a list of all of the web_test_metadata files in the specificied attributes.
+  """
+  metadata_files = []
+
+  for attr_name in attr_names:
+    if hasattr(ctx.attr, attr_name):
+      attr = getattr(ctx.attr, attr_name)
+      if lists.is_list_like(attr):
+        for item in attr:
+          _get_files_helper(metadata_files, item)
+      else:
+        _get_files_helper(metadata_files, attr)
+  return metadata_files
+
+
+def _get_files_helper(metadata_files, r):
+  if hasattr(r, 'transitive_web_test_metadata'):
+    lists.ensure_all_at_end_of_list(metadata_files,
+                                    r.transitive_web_test_metadata)
+  if hasattr(r, 'web_test_metadata'):
+    lists.ensure_all_at_end_of_list(metadata_files, r.web_test_metadata)
+
+
+def _web_test_files(named_files={}, archive_file=None):
+  return struct(archive_file=archive_file, named_files=named_files)
+
+
+def _web_test_files_to_string(ctx, files_struct, indent=''):
+  """Converts a web_test_files struct to a JSON string."""
+  result = indent + '{\n'
+  if files_struct.archive_file:
+    result += indent + '  "archiveFile": "' + files.runfiles_path(
+        ctx, files_struct.archive_file) + '",\n'
+  result += indent + '  "namedFiles": {'
+  first = True
+  for n, f in files_struct.named_files.items():
+    if first:
+      first = False
+    else:
+      result += ','
+    result += '\n' + indent + '    "' + n + '": "'
+    if type(f) == type(''):
+      result += f
+    else:
+      result += files.runfiles_path(ctx, f)
+    result += '"'
+  result += '\n' + indent + '  }\n' + indent + '}'
+  return result
+
+
+def _aspect_impl(target, ctx):
+  transitive_files = _get_files(ctx.rule, ['data', 'deps', 'srcs'])
+
+  if hasattr(target, 'web_test_metadata'):
+    for m in target.web_test_metadata:
+      lists.ensure_at_end_of_list(transitive_files, m)
+
+  return struct(transitive_web_test_metadata=transitive_files)
+
+
+_aspect = aspect(
+    implementation=_aspect_impl,
+    attr_aspects=['deps', 'data', 'srcs', 'browser', 'config'],)
```

Comment:
I am not confident that I want to use aspects; which is part of why I added you to this code review. But consider the following rules:

```bazel
web_test_named_file(
    name = "chrome",
    file = "chrome.exe",
)

web_test_named_file(
    name = "chromedriver",
    file = "chromedriver.exe",
)

browser(
    name = "chrome-windows",
    data = [
        ":chrome",
        ":chromedriver",
    ]
)
```

This will result in a metadata file with the following:

```json
{
    "webTestFiles": [
        {
            "namedFiles": {
                "chrome": "path/to/chrome.exe",
                "chromedriver": "path/to/chromedriver.exe"
            }   
        }
    ]
}
```
However without aspects, if this changes to:
```bazel
web_test_named_file(
    name = "chrome",
    file = "chrome.exe",
)

web_test_named_file(
    name = "chromedriver",
    file = "chromedriver.exe",
)

filegroup(
    name = "chrome_fg",
    data = [
        ":chrome",
        ":chromedriver",
    ]
)

browser(
    name = "chrome-windows",
    data = [
        ":chrome_fg",
    ]
)
```
The generated metadata file will be empty as it will have lost the connection to the web_test_metadata providers from the web_test_named_file rules. Using the aspect solved this problem.

That said, I am not fully convinced that this benefit is worth the cost. 

First, it seems unlikely that somebody would structure their rules this way. 

Second, it actually makes it harder to inspect what is happening at each step in the dependency graph. With the current rules, every web_test_* and browser target generates a single merged metadata file that can be examined that is the merger of all of the incoming metadata files. With the aspects, each target just generates it own little metadata fragments, and no merging occurs until we get to a web_test rule.

Given these tradeoffs, what are your thoughts on the aspect versus non-aspect version?